kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java

```diff
@@ -36,23 +40,34 @@ public class SnapshotImpl implements Snapshot {
     private final long version;
     private final LogReplay logReplay;
     private final Lazy<Tuple2<Protocol, Metadata>> protocolAndMetadata;
+    private final Optional<SnapshotHint> snapshotHint;
 
     public SnapshotImpl(
-        Path logPath,
-        Path dataPath,
-        long version,
-        LogSegment logSegment,
-        TableClient tableClient,
-        long timestamp) {
+            Path logPath,
+            Path dataPath,
+            long version,
+            LogSegment logSegment,
+            TableClient tableClient,
+            long timestamp,
+            SnapshotManager snapshotManager,
+            Optional<SnapshotHint> snapshotHint) {
         this.dataPath = dataPath;
         this.version = version;
-
         this.logReplay = new LogReplay(
             logPath,
             dataPath,
             tableClient,
-            logSegment);
-        this.protocolAndMetadata = new Lazy<>(logReplay::loadProtocolAndMetadata);
+            logSegment,
+            snapshotHint);
+        this.protocolAndMetadata = new Lazy<>(() -> {
+            // Construct the SnapshotHint lazily. i.e. only create it when some caller/consumer of
+            // this SnapshotImpl instance decides to load the protocol and metadata.
```

Comment:
In Delta spark, the metadata and protocol are needed ~immediately (**) and so cannot really be lazy. Is that not the case here?

(**) i.e. query planner triggered the snapshot because it needed the table schema

Reply:
https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/Snapshot.scala#L169

We are not doing this yet. Not sure why. @vkorukanti ?

Reply:
We [validate](https://github.com/delta-io/delta/blob/master/kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java#L179) as part of the loading the protocol and metadata. 

If I remember correctly, we want the cost of asking for a snapshot should be quick and not load something unless until is asked. In the case of the query planner, it will be asking for `Snapshot.getSchema` so that should trigger the loading.

We can change to follow the Delta-Spark model.

Reply:
We already had to perform I/O to create the snapshot at all (LIST) so I don't see much benefit in being lazy about P&M. Especially given the complexity and code tangling it introduces.